test/threadstest.c

```diff
@@ -184,10 +187,279 @@ static int test_thread_local(void)
+        return 0;
+
+    return 1;
+}
+
+static int test_thread_spawn(void)
+{
+    uint32_t local;
+    uint32_t retval;
+    void *task;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_enabled(NULL), 0))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_enable(NULL, 1), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_cap(NULL, 2), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), 2))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_cap(NULL, -1), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), -1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_cap(NULL, 1), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_spawn_worker(NULL, NULL), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), 2))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_cap(NULL, -1), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), -1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_cap(NULL, 1), 1))
+        return 0;
+
+    /* this will test a worker */
+
+    local = 2;
+    retval = 0;
+    task = CRYPTO_THREAD_start(NULL, test_thread_native_fn, &local);
+    if (!TEST_ptr(task))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_join(NULL, task, &retval),1))
+        return 0;
+
+    if (!TEST_int_eq(retval, 2) || !TEST_int_eq(fn1_glob, local))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_clean(NULL,task),1))
+        return 0;
+
+    /* this will test an internally spawned thread */
+
+    local = 3;
+    retval = 0;
+    task = CRYPTO_THREAD_start(NULL, test_thread_native_fn, &local);
+    if (!TEST_ptr(task))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), 0))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_join(NULL, task, &retval),1))
+        return 0;
+
+    if (!TEST_int_eq(retval, 2) || !TEST_int_eq(fn1_glob, local))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_clean(NULL,task),1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_disable(NULL), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_enabled(NULL), 0))
+        return 0;
+
+    return 1;
+}
+
+static int test_thread_spawn_policy(void)
+{
+    uint32_t local;
+    uint32_t retval;
+    void *task1, *task2, *task3;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_enabled(NULL), 0))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_enable(NULL, 2), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_get_available_threads(NULL), 2))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_spawn_worker(NULL, NULL), 1))
+        return 0;
+
+    if (!TEST_int_eq(CRYPTO_THREAD_spawn_worker(NULL, NULL), 1))
```

Comment:
I don't see anywhere that we test the callback argument. You might also want to test that, where you have two OPENSSL_CTXs, settings in one do not affect the other.

Reply:
Resolved: I've added both a test case using the callback argument and incorporated another OPENSSL_CTX into some of the existing tests: https://github.com/ckalina/openssl/commit/c7142e1603c4a6b2208d8ddfc190d6d5a178a4f9